cpp/src/arrow/dataset/file_base.cc

```diff
@@ -47,77 +46,161 @@ Result<ScanTaskIterator> FileDataFragment::Scan(ScanContextPtr context) {
-  auto file_to_fragment = [options, this](std::unique_ptr<fs::FileStats> stats,
-                                          std::shared_ptr<DataFragment>* out) {
-    FileSource src(stats->path(), filesystem_.get());
-    ARROW_ASSIGN_OR_RAISE(*out, format_->MakeFragment(src, options));
+  DCHECK_OK(forest_.Visit([&](fs::PathForest::Ref ref) {
+    repr += "\n" + ref.stats().path();
+
+    if (!partitions_[ref.i]->Equals(true)) {
+      repr += ": " + partitions_[ref.i]->ToString();
+    }
+
     return Status::OK();
-  };
+  }));
 
-  return MakeMaybeMapIterator(file_to_fragment, std::move(file_it));
+  return repr;
 }
 
-bool FileSystemDataSource::PartitionMatches(const fs::FileStats& stats,
-                                            ExpressionPtr filter) {
-  if (filter == nullptr) {
-    return true;
+util::optional<std::pair<std::string, std::shared_ptr<Scalar>>> GetKey(
+    const Expression& expr) {
+  if (expr.type() != ExpressionType::COMPARISON) {
+    return util::nullopt;
   }
 
-  auto found = partitions_.find(stats.path());
-  if (found == partitions_.end()) {
-    // No partition attached to current node (directory or file), continue.
-    return true;
+  const auto& cmp = internal::checked_cast<const ComparisonExpression&>(expr);
+  if (cmp.op() != compute::CompareOperator::EQUAL) {
+    return util::nullopt;
   }
 
-  auto expr = found->second->Assume(*filter);
-  if (expr->IsNull() || expr->Equals(false)) {
-    // selector is not satisfiable; don't recurse in this branch.
-    return false;
+  // TODO(bkietz) allow this ordering to be flipped
+
+  if (cmp.left_operand()->type() != ExpressionType::FIELD) {
+    return util::nullopt;
   }
 
-  return true;
+  if (cmp.right_operand()->type() != ExpressionType::SCALAR) {
+    return util::nullopt;
+  }
+
+  return std::make_pair(
+      internal::checked_cast<const FieldExpression&>(*cmp.left_operand()).name(),
+      internal::checked_cast<const ScalarExpression&>(*cmp.right_operand()).value());
+}
+
+struct SourceAndOptions {
+  FileSource source;
+  ScanOptionsPtr options;
+
+  // required by optional
+  bool operator==(const SourceAndOptions& other) const { return this == &other; }
+};
+
+DataFragmentIterator FileSystemDataSource::GetFragmentsImpl(ScanOptionsPtr options) {
+  std::vector<SourceAndOptions> files;
+
+  auto collect_files = [&](fs::PathForest::Ref ref) -> fs::PathForest::MaybePrune {
+    auto partition = partitions_[ref.i];
+    auto expr = options->filter->Assume(partition);
+
+    if (expr->IsNull() || expr->Equals(false)) {
+      // directories (and descendants) which can't satisfy the filter are pruned
+      return fs::PathForest::Prune;
+    }
+
+    if (ref.stats().IsFile()) {
+      SourceAndOptions file = {FileSource(ref.stats().path(), filesystem_.get()),
+                               std::make_shared<ScanOptions>(*options)};
+
+      // use simplified filter
+      for (auto ancestor = ref.parent(); ancestor.forest == &forest_;
```

Comment:
Storing parent partition information in children also leads to `O(n^2)/O(n lg n)` performance since simplification against a deep child's partition information will still require simplification against the `O(depth)` sub expressions which correspond to its parents.

Storing exclusively the current segment's partition information reduces the amount of destructuring logic required to extract partition key values, where available. It would be possible to write an arbitrary expression visitor to accumulate key value pairs given an arbitrary expression (for example, `"i"_ == 3 and "f"_ > 1.0 and "s"_ == "yo"` -> `{{"i", 3},{"s","yo"}}`), but that seems unacceptably brittle.

Beyond that, IMHO storing parent partition information is redundant and confusing during debugging.

Reply:
I've removed the ancestor crawl.